File: Leetcode/435/solution.py
# https://leetcode.com/problems/non-overlapping-intervals/description/
class Solution(object):
    # O(nlogn) runtime, O(1) space
    def eraseOverlapIntervals(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        """
        if len(intervals) == 0:
            return 0
        intervals.sort(key=lambda interval: interval.start)
        prev, count = 0, 0
        for i in range(1, len(intervals)):
            if intervals[prev].end > intervals[i].start:
                if intervals[prev].end > intervals[i].end:
                    prev = i
                count += 1
            else:
                prev = i
        return count

File: Leetcode/435/solution.java
// https://leetcode.com/problems/non-overlapping-intervals/description/
public class Solution {
    class myComparator implements Comparator<Interval> {
        public int compare(Interval a, Interval b) {
            return a.start - b.start;
        }
    }

    // O(nlogn) runtime, O(1) space
    public int eraseOverlapIntervals(Interval[] intervals) {
        if (intervals.length == 0) {
            return 0;
        }
        Arrays.sort(intervals, new myComparator());
        int prev = 0, count = 0;
        for (int i = 1; i < intervals.length; i++) {
            if (intervals[prev].end > intervals[i].start) {
                if (intervals[prev].end > intervals[i].end) {
                    prev = i;
                }
                count++;
            } else {
                prev = i;
            }
        }
        return count;
    }
}
